include/ginkgo/core/base/executor.hpp

```diff
@@ -1090,6 +1167,19 @@ class CudaExecutor : public detail::ExecutorBase<CudaExecutor>,
 
     GKO_ENABLE_FOR_ALL_EXECUTORS(GKO_OVERRIDE_RAW_COPY_TO);
 
+    GKO_DEFAULT_OVERRIDE_VERIFY_MEMORY(OmpExecutor, false);
+
+    GKO_DEFAULT_OVERRIDE_VERIFY_MEMORY(ReferenceExecutor, false);
+
+    GKO_DEFAULT_OVERRIDE_VERIFY_MEMORY(DpcppExecutor, false);
+
+    bool verify_memory_to(const HipExecutor *dest_exec) const override;
+
+    bool verify_memory_to(const CudaExecutor *dest_exec) const override
+    {
+        return device_id_ == dest_exec->get_device_id();
+    }
```

Comment:
Why is this in the header file, but not the corresponding HIP function?

Reply:
The issue is a chicken and egg problem. Who came first, CUDA or HIP? at most, I could do it for one of them, but not for both, so I decided to do it for none instead.

In more details, we have a predeclaration of all classes to allow for friendships. In order, CUDA Executor is declared then HIP. At the point in time where we want to call, inside the CudaExecutor, `HIP->get_device_id()`, we still only have incomplete information about HIP, and thus cannot find that function.

Reply:
Could we define both in the source files instead?

Reply:
You mean also the Cuda->Cuda one? The annoying thing is we then also need to implement the `device_hook`, but I can do that, yes.

Reply:
I've done that by moving the two implementations into `devices/{cuda|hip}/executor.cpp`, I'm not sure if it's more or less confusing, but at least I don't have to duplicate the implementation this way.